From 063236dfd55da750b257603de502f8dbdcc59789 Mon Sep 17 00:00:00 2001
From: Eric Kilmer <eric.d.kilmer@gmail.com>
Date: Sat, 1 Jan 2022 17:02:51 -0500
Subject: [PATCH] [cmake] Fix interface settings on exported LLVM targets

This makes it easier for users to link to LLVM in their CMake projects
without having to redo the logic required for setting the correct
properties like include path, compiler flags and options, etc.
---
 llvm/cmake/modules/AddLLVM.cmake | 17 +++++++++++++++--
 1 file changed, 15 insertions(+), 2 deletions(-)

diff --git a/llvm/cmake/modules/AddLLVM.cmake b/llvm/cmake/modules/AddLLVM.cmake
index 5a9480aa6768..2ff0fa07f89c 100644
--- a/llvm/cmake/modules/AddLLVM.cmake
+++ b/llvm/cmake/modules/AddLLVM.cmake
@@ -9,7 +9,7 @@ function(llvm_update_compile_flags name)
     set(update_src_props ON)
   endif()
 
-  list(APPEND LLVM_COMPILE_CFLAGS " ${LLVM_COMPILE_FLAGS}")
+  list(APPEND LLVM_COMPILE_CFLAGS "${LLVM_COMPILE_FLAGS}")
 
   # LLVM_REQUIRES_EH is an internal flag that individual targets can use to
   # force EH
@@ -59,25 +59,37 @@ function(llvm_update_compile_flags name)
   string(REPLACE ";" " " target_compile_flags " ${LLVM_COMPILE_FLAGS}")
   string(REPLACE ";" " " target_compile_cflags " ${LLVM_COMPILE_CFLAGS}")
 
+  set_property(TARGET ${name} APPEND PROPERTY INTERFACE_COMPILE_OPTIONS ${LLVM_COMPILE_CFLAGS})
   if(update_src_props)
+    set(cpp_target FALSE)
     foreach(fn ${sources})
       get_filename_component(suf ${fn} EXT)
       if("${suf}" STREQUAL ".cpp")
         set_property(SOURCE ${fn} APPEND_STRING PROPERTY
           COMPILE_FLAGS "${target_compile_flags}")
+        set(cpp_target TRUE)
       endif()
       if("${suf}" STREQUAL ".c")
         set_property(SOURCE ${fn} APPEND_STRING PROPERTY
           COMPILE_FLAGS "${target_compile_cflags}")
       endif()
     endforeach()
+    if(cpp_target)
+      set_property(TARGET ${name} APPEND PROPERTY INTERFACE_COMPILE_OPTIONS ${LLVM_COMPILE_FLAGS})
+    endif()
   else()
     # Update target props, since all sources are C++.
     set_property(TARGET ${name} APPEND_STRING PROPERTY
       COMPILE_FLAGS "${target_compile_flags}")
+    set_property(TARGET ${name} APPEND PROPERTY INTERFACE_COMPILE_OPTIONS ${LLVM_COMPILE_FLAGS})
   endif()
 
   set_property(TARGET ${name} APPEND PROPERTY COMPILE_DEFINITIONS ${LLVM_COMPILE_DEFINITIONS})
+  set_property(TARGET ${name} APPEND PROPERTY INTERFACE_COMPILE_DEFINITIONS ${LLVM_COMPILE_DEFINITIONS})
+  target_include_directories(${name} PUBLIC
+    $<BUILD_INTERFACE:${LLVM_INCLUDE_DIR}>
+    $<BUILD_INTERFACE:${LLVM_MAIN_INCLUDE_DIR}>
+  )
 endfunction()
 
 function(add_llvm_symbol_exports target_name export_file)
@@ -824,7 +836,8 @@ macro(add_llvm_library name)
               ${export_to_llvmexports}
               LIBRARY DESTINATION lib${LLVM_LIBDIR_SUFFIX} COMPONENT ${name}
               ARCHIVE DESTINATION lib${LLVM_LIBDIR_SUFFIX} COMPONENT ${name}
-              RUNTIME DESTINATION bin COMPONENT ${name})
+              RUNTIME DESTINATION bin COMPONENT ${name}
+              INCLUDES DESTINATION include)
 
       if (NOT LLVM_ENABLE_IDE)
         add_llvm_install_targets(install-${name}
-- 
2.33.1

